aya/src/maps/mod.rs

```diff
@@ -604,28 +622,32 @@ impl MapData {
     /// Returns the file descriptor of the map.
     ///
     /// Can be converted to [`RawFd`] using [`AsRawFd`].
-    pub fn fd(&self) -> Option<MapFd> {
-        self.fd.map(MapFd)
+    pub fn fd(&self) -> Option<MapFd<'_>> {
+        self.fd.as_ref().map(|fd| MapFd(fd.as_fd()))
     }
 }
 
-impl Drop for MapData {
-    fn drop(&mut self) {
-        // TODO: Replace this with an OwnedFd once that is stabilized.
-        if let Some(fd) = self.fd.take() {
-            unsafe { libc::close(fd) };
-        }
+impl MapData {
+    /// Attempts to duplicate MapData
+    ///
+    /// Fails if if the inner file descriptor could not be cloned
+    // TODO (AM): should we return MapError? New variant? Maybe make MapError non_exhaustive?
+    pub fn try_clone(&self) -> io::Result<Self> {
+        let fd = self.fd.as_ref().map(|f| f.try_clone()).transpose()?;
+
+        Ok(MapData {
+            fd,
+            obj: self.obj.clone(),
+            btf_fd: self.btf_fd,
+            pinned: self.pinned,
+        })
     }
 }
 
+// TODO (AM): DELETE?
 impl Clone for MapData {
     fn clone(&self) -> MapData {
-        MapData {
-            obj: self.obj.clone(),
-            fd: self.fd.map(|fd| unsafe { libc::dup(fd) }),
-            btf_fd: self.btf_fd,
-            pinned: self.pinned,
-        }
+        self.try_clone().unwrap()
```

Comment:
I think we should delete this as implicit panics on a clone is weird and clone can definitely fail but I left it here so the choice is up to the maintainers on what feels best here.